### PR TITLE
Add an option to center children around the new parent when reparenting

### LINKED ISSUE #8449: New node created from "Reparent to New Node" positioned relative to the child nodes, not the origin
### Describe the project you are working on

A 3D Platformer with Collectibles (Coins, Health Pickups, etc.)

### Describe the problem or limitation you are having in your project

Selecting a group of collectibles and reparenting to new node created a 3D Node at (0, 0, 0). I wanted to create a reusable scene that's a circle of these collectibles, but placing the node at the origin rather than somewhere reasonably close to the original nodes makes rotation and any other transformations behave odd.

Similarly, if I have a mesh in a level that I want to add collision to, it would make more sense to be able to select the mesh node and reparent to a StaticBody3D, but with the current flow, this creates the StaticBody3D at the origin, so you would need to add a StaticBody3D as a child of the mesh, then unparent the StaticBody3D, and finally parent the mesh to the StaticBody3D.

For an example of the collectible situation, here is a ring of coins that I put in my level, just to see if it's something I would like:
image

Then, I decide that I want these to spin around a common point, and I may want to reuse that for different parts of the level. So, I select to reparent them a new node, and select Node3D. This currently creates a node centered about (0, 0, 0):
image

What would be much more preferable is a node in the center of the coins, or even a node placed at the first coin. In that case it would be at least easy to adjust the positions of the coins to fix the issue.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Placing this new parent node at a more reasonable position like the center of the selected node(s) would solve both of these situations and produce more expected behavior than creating a node at the center of the world.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Add a setting or change the implementation to allow this new node to be centered around the average position of the reparented node(s).

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This can be worked around with a plugin, but the time lost trying to fix it each time is so small that I don't believe most users would go out of their way to find a plugin to solve this.

### Is there a reason why this should be core and not an add-on in the asset library?

This behavior seems much more intuitive to what people use the "Reparent to New Node" feature for than creating a seemingly unrelated node. And as stated before, each time someone would run into this issue, it would probably only cost about a minute or two to solve, but level designers would run into this situation quite a few times throughout the creation of a game.

### PR DESCRIPTION
**Pull request by Bitlytic**
Adds the functionality of proposal [#8449](https://github.com/godotengine/godot-proposals/issues/8449), in the case that others feel this would be a good enhancement.

This enhances the "Reparent to New Node" tool in the Scene view by adding the ability to create a new based around the selected node(s) rather than always adding the new parent node at the origin. 

_Production edit: Closes https://github.com/godotengine/godot-proposals/issues/8449_

### CODE REVIEW

File: editor/editor_settings.cpp [hunk 1]
```diff
@@ -527,6 +527,7 @@ void EditorSettings::_load_defaults(Ref<ConfigFile> p_extra_config) {
 	// SceneTree
 	_initial_set("docks/scene_tree/start_create_dialog_fully_expanded", false);
 	_initial_set("docks/scene_tree/auto_expand_to_selected", true);
+	_initial_set("docks/scene_tree/center_node_on_reparent", true);
```
Comment (AThousandShips) [inline editor/editor_settings.cpp:530]:
```suggestion
	_initial_set("docks/scene_tree/center_node_on_reparent", false);
```
New settings that change behaviour should be turned off by default, they should be opt-in

File: editor/scene_tree_dock.cpp [hunk 2]
```diff
@@ -2580,6 +2590,44 @@ void SceneTreeDock::_create() {
 
 		// This works because editor_selection was cleared and populated with last created node in _do_create()
 		Node *last_created = editor_selection->get_selected_node_list().front()->get();
+
+		if (center_parent) {
+			// Find parent type and only average positions of relevant nodes
```
Comment (KoBeWi) [inline editor/scene_tree_dock.cpp:2595]:
```suggestion
			// Find parent type and only average positions of relevant nodes.
```

File: editor/scene_tree_dock.cpp [hunk 3]
```diff
@@ -2580,6 +2590,44 @@ void SceneTreeDock::_create() {
...
+			Node3D *parent_node_3d = Object::cast_to<Node3D>(last_created);
+			if (parent_node_3d) {
+				Vector3 position;
+				uint32_t node_count = 0;
+				for (Node *node : nodes) {
```
Comment (KoBeWi) [inline editor/scene_tree_dock.cpp:2600]:
```suggestion
				for (const Node *node : nodes) {
```
(if it's possible; you need to make the Node3D const too)

File: editor/scene_tree_dock.cpp [hunk 4]
```diff
@@ -2580,6 +2590,44 @@ void SceneTreeDock::_create() {
...
+					Node3D *node_3d = Object::cast_to<Node3D>(node);
+					if (node_3d) {
+						position += node_3d->get_global_position();
+						node_count++;
+					}
+				}
+
+				if (node_count > 0) {
+					parent_node_3d->set_global_position(position / (real_t)node_count);
```
Comment (KoBeWi) [inline editor/scene_tree_dock.cpp:2609]:
```suggestion
					parent_node_3d->set_global_position(position / node_count);
```
The cast is unnecessary (unless it silences some warning?)

**Review by KoBeWi — APPROVED** [review]
The implementation looks fine.

Though making it an editor setting disabled by default means poor discoverability. Maybe it could be an option in the `⋮` menu.

File: editor/scene_tree_dock.cpp [hunk 5]
```diff
@@ -2580,6 +2593,44 @@ void SceneTreeDock::_create() {
...
+					parent_node_3d->set_global_position(position / node_count);
+				}
+			}
+
+			Node2D *parent_node_2d = Object::cast_to<Node2D>(last_created);
+			if (parent_node_2d) {
+				Vector2 position;
+				uint32_t node_count = 0;
+				for (Node *node : nodes) {
+					Node2D *node_2d = Object::cast_to<Node2D>(node);
```
Comment (KoBeWi) [inline editor/scene_tree_dock.cpp:2621]:
```suggestion
				for (const Node *node : nodes) {
					const Node2D *node_2d = Object::cast_to<Node2D>(node);
```
Here too 🙃

File: editor/scene_tree_dock.cpp [hunk 6]
```diff
@@ -3287,6 +3338,9 @@ void SceneTreeDock::_update_tree_menu() {
 	tree_menu->add_check_item(TTR("Auto Expand to Selected"), TOOL_AUTO_EXPAND);
 	tree_menu->set_item_checked(tree_menu->get_item_index(TOOL_AUTO_EXPAND), EDITOR_GET("docks/scene_tree/auto_expand_to_selected"));
 
+	tree_menu->add_check_item(TTR("Center Node on Reparent"), TOOL_CENTER_PARENT);
```
Comment (KoBeWi) [inline editor/scene_tree_dock.cpp:3341]: I think this needs a tooltip explaining what it does, as it's not obvious.
Something like: `If enabled, Reparent to New Node option will create the new node in the center of the selected nodes, if  possible.`

File: editor/scene_tree_dock.cpp (hunk 5 above)
Comment (AThousandShips) [reply editor/scene_tree_dock.cpp:2621]: Please apply this suggestion

**Comment by akien-mga** [comment]
Pushed a rebase to squash the commits.

**Pull request merged**